### DISCUSSION TITLE
Snapshot becomes null when component properties are set by test

### DISCUSSION THREAD

**Discussion by domjanzsoo** [livewire/livewire#8602] [1. Help]
I have a form with some text fields and a file uploader. All works well, but I'm unable to write tests for the upload field.
Keep getting the following error:

![image](https://github.com/livewire/livewire/assets/40408231/c95697fb-c6be-49a3-b2ab-c9a0b482e549)

This is my component:

    <?php

    namespace App\Livewire\Users;

    use Livewire\Component;
    use Illuminate\Support\Facades\Hash;
    use App\Contract\UserRepositoryInterface;
    use App\Contract\PermissionRepositoryInterface;
    use App\Contract\RoleRepositoryInterface;
    use Exception;
    use Illuminate\Auth\Access\AuthorizationException;
    use Livewire\WithFileUploads;

    class Add extends Component
    {
    use WithFileUploads;

    private $userRepository;
    private $permissionRepository;
    private $roleRepository;

    public array $state = [
        'full_name'             => null,
        'email'                 => null,
        'password'              => null,
        'password_confirmation' => null,
        'verified'              => false,
        'profile_picture'       => null,
        'permissions'           => [],
        'roles'                 => []
    ];

    protected $rules = [
        'state.full_name'               => 'required',
        'state.email'                   => 'required|email|unique:users,email',
        'state.password'                => 'required|confirmed|min:6',
        'state.password_confirmation'   => 'required',
        'state.profile_picture'         => 'image|max:2048|nullable',
        'state.permissions'             => 'array',
        'state.roles'                   => 'array'
        
    ];

    protected $listeners = [
        'user-permissions' => 'handlePermissions'
    ];
    
    public function messages(): array
    {
        return [
            'state.full_name.required' => trans('validation.required', ['attribute' => 'name']),
            'state.email.required' => trans('validation.required', ['attribute' => 'email']),
            'state.email.unique' => trans('validation.unique', ['attribute' => 'user email']),
            'state.email.email' => trans('validation.email', ['attribute' => 'user email']),
            'state.password.required' => trans('validation.required', ['attribute' => 'password']),
            'state.password.min' => trans('validation.min.string', ['attribute' => 'password', 'min' => 6]),
            'state.password.confirmed' => trans('validation.confirmed', ['attribute' => 'password']),
            'state.profile_picture.image' => trans('validation.image', ['attribute' => 'profile picture']),
            'state.profile_picture.max' => trans('validation.max.file', ['max' => '2048', 'attribute' => 'profile picture'])
        ];
    }

    // public function updatedStatePassword(): void
    // {
    //     $this->validateOnly('state.password');
    // }

    // public function updatedStatePasswordConfirmation(): void
    // {
    //     $this->validateOnly('state.password');
    // }

    public function handlePermissions(array $selections): void
    {
        $this->state['permissions'] = [];

        foreach ($selections as $id => $selection) {
            if ($selection['selected'] && !in_array($id, $this->state['permissions'])) {
                array_push($this->state['permissions'], $id);
            }
        }
    }

    public function handleRoles(array $selections): void
    {
        $this->state['roles'] = [];

        foreach ($selections as $id => $selection) {
            if ($selection['selected'] && !in_array($id, $this->state['roles'])) {
                array_push($this->state['roles'], $id);
            }
        }
    }

    public function render()
    {
        return view('livewire.users.add', [
            'permissions'   => $this->permissionRepository->getAll(),
            'roles'         => $this->roleRepository->getAll()
        ]);
    }

    public function boot(
        UserRepositoryInterface $userRepository,
        PermissionRepositoryInterface $permissionRepository,
        RoleRepositoryInterface $roleRepository
    )
    {
        $this->userRepository = $userRepository;
        $this->permissionRepository = $permissionRepository;
        $this->roleRepository = $roleRepository;
    }

    public function addUser(): void
    {     
        if (!access_control()->canAccess(auth()->user(), 'add_user')) {
            throw new AuthorizationException(trans('errors.unauthorized_action', ['action' => 'add user']));
        }

        $validatedData = $this->validate();

        // dd($validatedData);

        try {
            $user = $this->userRepository->createUser(
                [
                    'name'      => $validatedData['state']['full_name'],
                    'email'     => $validatedData['state']['email'],
                    'password'  => Hash::make($validatedData['state']['password'])
                ],
                $validatedData['state']['permissions'],
                $validatedData['state']['roles']
            );

            if ($validatedData['state']['profile_picture']) {
                $profilePictureFileName = md5($user->id) . '.' . $validatedData['state']['profile_picture']->extension();
    
                $validatedData['state']['profile_picture']->storeAs('/avatar', $profilePictureFileName, $disk = config('filesystems.default'));
    
                $user->profile_photo_path = 'storage/avatar/' . $profilePictureFileName;
                $user->save();
            }

        } catch(Exception $exception) {
            $this->dispatch('toastr', ['type' => 'error', 'message' => $exception->getMessage()]);
        }

        $this->state['full_name'] = null;
        $this->state['email'] = null;
        $this->state['password'] = null;
        $this->state['password_confirmation'] = null;
        $this->state['profile_picture'] = null;
        $this->state['permissions'] = [];
        $this->state['roles'] = [];

        $this->dispatch('toastr', ['type' => 'confirm', 'message' => trans('notifications.successfull_creation', ['entity' => 'User'])]);
        $this->dispatch('user-permissions-submitted');
        $this->dispatch('user-roles-submitted');

        return;
    }
    }

and this is the test of the file upload

    /** @test */
    public function fails_with_oversized_profile_image()
    {
        $this->actingAs($this->userWithAddPermissionAccess);

        $this->assertEquals(3, User::count());

        Storage::fake('public');

        $profileImage = UploadedFile::fake()->create('blaa.jpg', 145412);
        
        Livewire::test(CreateUser::class)
            ->set('state', [
                'full_name' => 'Joe Doe',
                'email' => 'test.user@email.com',
                'password' => 'password',
                'password_confirmation' => 'password',
                'profile_picture' => $profileImage,
                'permissions' => [],
                'roles' => []
            ])
            ->call('addUser')
            ->assertHasErrors(['state.profile_picture' => 'max']);

        $this->assertEquals(3, User::count());
    }

After debugging I concluded, that the following happens: 
- at first render the fields are set to their initial values and both boot and render lifecycle methods run successfully, and that upload method of HandleComponents receives the right snapshot
- once all the properties set with the UploadedFile object included, the component re-runs the lifecycle methods and the snapshot passed to HandleComponents becomes null, which returns the error attached above.

without file the test runs with no problem.

My question is: When is that update method called? What is the purpose of it? Why the snapshot becomes null, when the file upload is included?

Thanks.

**Comment by lrljoe** [comment]
I mean, first off, I'd really recommend that you use a Livewire Form, not an array of properties, it makes life much simpler!

Use the fake image() method to create an image, rather than just a file.

Separate your file upload out of your array, test purely the file upload section.

Keep in mind how file uploads work, the Livewire 2 site has a working example of testing a file upload, start with that, and slowly add what you need.

**Reply by domjanzsoo** [reply]
the Livewire2 site's example does not consider the usage of the temporaryUrl() method of the FileUpload object. That's why that example does not help me. Also that example not giving you any validation hint, it's simply checking that the upload was successful. By using the image() method of the faker I won't be able to set the size and the mime type of the file (that's possible only with the create() method, I tried to add those properties to the faked object, but it doesn't work). 
The use of Livewire Form is a good point though, I'm going to check it (hopefully that won't cause any more testing problems).
Thanks.

**Comment by itsmesaskovic** [comment]
> I have a form with some text fields and a file uploader. All works well, but I'm unable to write tests for the upload field. Keep getting the following error:
> 
> ![image](https://private-user-images.githubusercontent.com/40408231/342753748-c95697fb-c6be-49a3-b2ab-c9a0b482e549.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEzNjAyMTIsIm5iZiI6MTc0MTM1OTkxMiwicGF0aCI6Ii80MDQwODIzMS8zNDI3NTM3NDgtYzk1Njk3ZmItYzZiZS00OWEzLWIyYWItYzlhMGI0ODJlNTQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA3VDE1MDUxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMWQ0NTQ0MWQ2NzVlZmI1MGExMTcxZTY0NTcxMmIxODAxMjVjZDhlN2JjMTNmMTA1Nzg3MWRjZmZjOTMzMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e55sGQgepOdrlqf2KnnGRc_A6qpacHl2CFmk3nCLj7I)
> 

Hi @domjanzsoo, this is a while back, but over the last few days I ran into similar issues and thought I'd share what fixed it on our side. You can read more about it here: [https://github.com/livewire/livewire/discussions/9225](url)
